src/crystal/syntax_highlighter/colorize.cr

```diff
@@ -0,0 +1,82 @@
+require "colorize"
+require "../syntax_highlighter"
+
+# A syntax highlighter that renders Crystal source code with ANSI escape codes
+# suitable for terminal highlighting.
+#
+# ```
+# code = %(foo = bar("baz\#{PI + 1}") # comment)
+# colorized = Crystal::SyntaxHighlighter::Colorize.highlight(code)
+# colorized # => "foo \e[91m=\e[0m bar(\e[93m\"baz\#{\e[0;36mPI\e[0;93m \e[0;91m+\e[0;93m \e[0;35m1\e[0;93m}\"\e[0m) \e[90m# comment\e[0m"
+# ```
+class Crystal::SyntaxHighlighter::Colorize < Crystal::SyntaxHighlighter
+  # Highlights *code* and writes the result to *io*.
+  def self.highlight(io : IO, code : String)
+    new(io).highlight(code)
+  end
+
+  # Highlights *code* and returns the result.
+  def self.highlight(code : String)
+    String.build do |io|
+      highlight(io, code)
+    end
+  end
+
+  # Highlights *code* or returns unhighlighted *code* on error.
+  #
+  # Same as `.highlight(code : String)` except that any error is rescued and
+  # returns unhighlighted source code.
+  def self.highlight!(code : String)
+    highlight(code)
+  rescue
+    code
+  end
+
+  def initialize(@io : IO)
+  end
+
+  property colors : Hash(TokenType, ::Colorize::Color) = {
+    :comment           => :dark_gray,
+    :number            => :magenta,
+    :char              => :light_yellow,
+    :symbol            => :magenta,
+    :string            => :light_yellow,
+    :interpolation     => :light_yellow,
+    :const             => :cyan,
+    :operator          => :light_red,
+    :ident             => :light_green,
+    :keyword           => :light_red,
+    :primitive_literal => :magenta,
+    :self              => :blue,
+  } of TokenType => ::Colorize::Color
+
+  def render(type : TokenType, value : String)
+    colorize(type, value)
+  end
+
+  def render_delimiter(&)
+    ::Colorize.with.toggle(true).fore(colors[TokenType::STRING]).surround(@io) do
```

Comment:
Consider this suggestion: ensure the ability to use a consistent `colorize` object, e.g. to be able to use `.on_tty_only!`.

[`4405e53` (#11366)](https://github.com/crystal-lang/crystal/pull/11366/commits/4405e53fe2b50a30c8dd187e7fffbda5fcceddbf) should be reverted then

```suggestion
  def initialize(@io : IO, @colorize : ::Colorize::Object(String) = ::Colorize.with.toggle(true))
  end

  property colors : Hash(TokenType, ::Colorize::Color) = {
    :comment           => :dark_gray,
    :number            => :magenta,
    :char              => :light_yellow,
    :symbol            => :magenta,
    :string            => :light_yellow,
    :interpolation     => :light_yellow,
    :const             => :cyan,
    :operator          => :light_red,
    :ident             => :light_green,
    :keyword           => :light_red,
    :primitive_literal => :magenta,
    :self              => :blue,
  } of TokenType => ::Colorize::Color

  def render(type : TokenType, value : String)
    colorize(type, value)
  end

  def render_delimiter(&)
    @colorize.fore(colors[TokenType::STRING]).surround(@io) do
```